File: src/trading/hyperliquid_api.py
import asyncio
import logging
import aiohttp
from src.config_loader import CONFIG
from hyperliquid.exchange import Exchange
from hyperliquid.info import Info
from hyperliquid.utils import constants  # For MAINNET/TESTNET
from eth_account import Account
import logging

class HyperliquidAPI:
    def __init__(self):
        if "hyperliquid_private_key" in CONFIG and CONFIG["hyperliquid_private_key"]:
            self.wallet = Account.from_key(CONFIG["hyperliquid_private_key"])
        elif "mnemonic" in CONFIG and CONFIG["mnemonic"]:
            self.wallet = Account.from_mnemonic(CONFIG["mnemonic"])
        else:
            raise ValueError("Either HYPERLIQUID_PRIVATE_KEY/LIGHTER_PRIVATE_KEY or MNEMONIC must be provided")
        # Choose base URL: allow override via env-config; default MAINNET
        base_url = CONFIG.get("hyperliquid_base_url") or constants.MAINNET_API_URL
        self.info = Info(base_url)
        self.exchange = Exchange(self.wallet, base_url)

    def round_size(self, asset, amount):
        """Round amount to asset's szDecimals precision to avoid float_to_wire errors."""
        meta = self._meta_cache[0] if hasattr(self, '_meta_cache') and self._meta_cache else None
        if meta:
            universe = meta.get("universe", [])
            asset_info = next((u for u in universe if u.get("name") == asset), None)
            if asset_info:
                decimals = asset_info.get("szDecimals", 8)
                return round(amount, decimals)
        return round(amount, 8)

    async def place_buy_order(self, asset, amount, slippage=0.01):
        amount = self.round_size(asset, amount)
        return await asyncio.to_thread(self.exchange.market_open, asset, True, amount, None, slippage)

    async def place_sell_order(self, asset, amount, slippage=0.01):
        amount = self.round_size(asset, amount)
        return await asyncio.to_thread(self.exchange.market_open, asset, False, amount, None, slippage)

    async def place_take_profit(self, asset, is_buy, amount, tp_price):
        # TP as trigger order (market close at tp_price, reduce-only)
        amount = self.round_size(asset, amount)
        order_type = {"trigger": {"triggerPx": tp_price, "isMarket": True, "tpsl": "tp"}}
        return await asyncio.to_thread(self.exchange.order, asset, not is_buy, amount, tp_price, order_type, True)

    async def place_stop_loss(self, asset, is_buy, amount, sl_price):
        # SL as trigger order (market close at sl_price, reduce-only)
        amount = self.round_size(asset, amount)
        order_type = {"trigger": {"triggerPx": sl_price, "isMarket": True, "tpsl": "sl"}}
        return await asyncio.to_thread(self.exchange.order, asset, not is_buy, amount, sl_price, order_type, True)

    async def cancel_order(self, asset, oid):
        return await asyncio.to_thread(self.exchange.cancel, asset, oid)

    async def cancel_all_orders(self, asset):
        """Cancel all open orders for an asset."""
        try:
            open_orders = await asyncio.to_thread(self.info.frontend_open_orders, self.wallet.address)
            for order in open_orders:
                if order.get("coin") == asset:
                    oid = order.get("oid")
                    if oid:
                        await self.cancel_order(asset, oid)
            return {"status": "ok", "cancelled_count": len([o for o in open_orders if o.get("coin") == asset])}
        except Exception as e:
            logging.error(f"Cancel all orders error for {asset}: {e}")
            return {"status": "error", "message": str(e)}

    async def get_open_orders(self):
        """Return list of current open orders for this wallet."""
        try:
            return await asyncio.to_thread(self.info.frontend_open_orders, self.wallet.address)
        except Exception as e:
            logging.error(f"Get open orders error: {e}")
            return []

    def extract_oids(self, order_result):
        oids = []
        try:
            statuses = order_result["response"]["data"]["statuses"]
            for st in statuses:
                if "resting" in st and "oid" in st["resting"]:
                    oids.append(st["resting"]["oid"])
                if "filled" in st and "oid" in st["filled"]:
                    oids.append(st["filled"]["oid"])
        except Exception:
            pass
        return oids

    async def get_user_state(self):
        state = await asyncio.to_thread(self.info.user_state, self.wallet.address)
        positions = state.get("assetPositions", [])
        for pos_wrap in positions:
            pos = pos_wrap["position"]
            entry_px = float(pos.get("entryPx", 0) or 0)
            size = float(pos.get("szi", 0) or 0)
            side = "long" if size > 0 else "short"
            current_px = await self.get_current_price(pos["coin"]) if entry_px and size else 0.0
            pnl = (current_px - entry_px) * abs(size) if side == "long" else (entry_px - current_px) * abs(size)
            pos["pnl"] = pnl
        balance = float(state.get("withdrawable", 0.0))
        return {"balance": balance, "positions": [p["position"] for p in positions]}

    async def get_current_price(self, asset):
        mids = await asyncio.to_thread(self.info.all_mids)
        return float(mids.get(asset, 0.0))

    async def get_meta_and_ctxs(self):
        """Cache meta and asset contexts to avoid repeated calls."""
        if not hasattr(self, '_meta_cache') or not self._meta_cache:
            response = await asyncio.to_thread(self.info.meta_and_asset_ctxs)
            self._meta_cache = response
        return self._meta_cache

    async def get_open_interest(self, asset):
        try:
            data = await self.get_meta_and_ctxs()
            if isinstance(data, list) and len(data) >= 2:
                meta, asset_ctxs = data[0], data[1]
                universe = meta.get("universe", [])
                asset_idx = next((i for i, u in enumerate(universe) if u.get("name") == asset), None)
                if asset_idx is not None and asset_idx < len(asset_ctxs):
                    oi = asset_ctxs[asset_idx].get("openInterest")
                    return round(float(oi), 2) if oi else None
            return None
        except Exception as e:
            logging.error(f"OI fetch error for {asset}: {e}")
            return None

    async def get_funding_rate(self, asset):
        try:
            data = await self.get_meta_and_ctxs()
            if isinstance(data, list) and len(data) >= 2:
                meta, asset_ctxs = data[0], data[1]
                universe = meta.get("universe", [])
                asset_idx = next((i for i, u in enumerate(universe) if u.get("name") == asset), None)
                if asset_idx is not None and asset_idx < len(asset_ctxs):
                    funding = asset_ctxs[asset_idx].get("funding")
                    return round(float(funding), 8) if funding else None
            return None
        except Exception as e:
            logging.error(f"Funding fetch error for {asset}: {e}")
            return None

File: src/trading/hyperliquid-api.ts
import { Hyperliquid } from "hyperliquid";
import { Wallet } from "ethers";
import { settings } from "../config/settings.js";
import { logger } from "../config/logger.js";
import type { EnrichedPosition, FillEntry, OpenOrder, UserState } from "../types/index.js";

type MetaCache = [{ universe: Array<{ name: string; szDecimals?: number }> }, Array<{ openInterest?: string; funding?: string }>];

export class HyperliquidApi {
  private readonly sdk: Hyperliquid;
  private readonly walletAddress: string;
  private metaCache: MetaCache | null = null;
  private initialized = false;

  constructor() {
    if (!settings.hyperliquidPrivateKey) {
      throw new Error("Either HYPERLIQUID_PRIVATE_KEY/LIGHTER_PRIVATE_KEY or MNEMONIC must be provided");
    }
    this.walletAddress = new Wallet(settings.hyperliquidPrivateKey).address;
    this.sdk = new Hyperliquid({
      privateKey: settings.hyperliquidPrivateKey,
      testnet: settings.hyperliquidTestnet,
      walletAddress: this.walletAddress,
      enableWs: false,
    });
  }

  private async ensureInit(): Promise<void> {
    if (this.initialized) return;
    await this.sdk.ensureInitialized();
    this.initialized = true;
  }

  private async retry<T>(fn: () => Promise<T>, maxAttempts = 3, backoffBase = 500): Promise<T> {
    let lastError: unknown;
    for (let attempt = 0; attempt < maxAttempts; attempt++) {
      try {
        await this.ensureInit();
        return await fn();
      } catch (err) {
        lastError = err;
        logger.warn({ attempt: attempt + 1, maxAttempts, err }, "Hyperliquid call failed");
        if (attempt < maxAttempts - 1) {
          await new Promise((r) => setTimeout(r, backoffBase * 2 ** attempt));
        }
      }
    }
    throw lastError ?? new Error("Hyperliquid retry: unknown error");
  }

  private async getMetaAndCtxs(): Promise<MetaCache> {
    if (!this.metaCache) {
      this.metaCache = (await this.retry(() =>
        this.sdk.info.perpetuals.getMetaAndAssetCtxs()
      )) as MetaCache;
    }
    return this.metaCache;
  }

  roundSize(asset: string, amount: number): number {
    const meta = this.metaCache?.[0];
    const assetInfo = meta?.universe.find((u) => u.name === asset);
    const decimals = assetInfo?.szDecimals ?? 8;
    const factor = 10 ** decimals;
    return Math.round(amount * factor) / factor;
  }

  async placeBuyOrder(asset: string, amount: number, slippage = 0.01): Promise<unknown> {
    const size = this.roundSize(asset, amount);
    return this.retry(() => this.sdk.custom.marketOpen(asset, true, size, undefined, slippage));
  }

  async placeSellOrder(asset: string, amount: number, slippage = 0.01): Promise<unknown> {
    const size = this.roundSize(asset, amount);
    return this.retry(() => this.sdk.custom.marketOpen(asset, false, size, undefined, slippage));
  }

  async placeTakeProfit(asset: string, isBuy: boolean, amount: number, tpPrice: number): Promise<unknown> {
    const size = this.roundSize(asset, amount);
    return this.retry(() =>
      this.sdk.exchange.placeOrder({
        coin: asset,
        is_buy: !isBuy,
        sz: size,
        limit_px: tpPrice,
        order_type: { trigger: { triggerPx: tpPrice, isMarket: true, tpsl: "tp" } },
        reduce_only: true,
      })
    );
  }

  async placeStopLoss(asset: string, isBuy: boolean, amount: number, slPrice: number): Promise<unknown> {
    const size = this.roundSize(asset, amount);
    return this.retry(() =>
      this.sdk.exchange.placeOrder({
        coin: asset,
        is_buy: !isBuy,
        sz: size,
        limit_px: slPrice,
        order_type: { trigger: { triggerPx: slPrice, isMarket: true, tpsl: "sl" } },
        reduce_only: true,
      })
    );
  }

  extractOids(orderResult: unknown): number[] {
    const oids: number[] = [];
    try {
      const result = orderResult as {
        response?: { data?: { statuses?: Array<Record<string, { oid?: number }>> } };
      };
      for (const st of result.response?.data?.statuses ?? []) {
        if (st.resting?.oid !== undefined) oids.push(st.resting.oid);
        if (st.filled?.oid !== undefined) oids.push(st.filled.oid);
      }
    } catch {
      // ignore malformed responses
    }
    return oids;
  }

  async getOpenOrders(): Promise<OpenOrder[]> {
    try {
      const orders = await this.retry(() =>
        this.sdk.info.getFrontendOpenOrders(this.walletAddress)
      );
      return (orders as Array<Record<string, unknown>>).map((o) => {
        const order: OpenOrder = {
          coin: o.coin as string,
          oid: o.oid as number,
          isBuy: o.side === "B" || o.isBuy === true,
          sz: parseFloat(String(o.sz ?? 0)),
          px: parseFloat(String(o.limitPx ?? o.px ?? 0)),
        };
        const ot = o.orderType as { trigger?: { triggerPx?: string } } | undefined;
        if (ot?.trigger?.triggerPx) {
          order.triggerPx = parseFloat(ot.trigger.triggerPx);
        }
        return order;
      });
    } catch (err) {
      logger.error({ err }, "Get open orders error");
      return [];
    }
  }

  async getRecentFills(limit = 50): Promise<FillEntry[]> {
    try {
      const fills = await this.retry(() => this.sdk.info.getUserFills(this.walletAddress));
      const list = Array.isArray(fills) ? fills : [];
      return list.slice(-limit) as FillEntry[];
    } catch (err) {
      logger.error({ err }, "Get recent fills error");
      return [];
    }
  }

  async getCurrentPrice(asset: string): Promise<number> {
    const mids = await this.retry(() => this.sdk.info.getAllMids());
    return parseFloat(String((mids as Record<string, string | number>)[asset] ?? 0));
  }

  async getUserState(): Promise<UserState> {
    const raw = await this.retry(() =>
      this.sdk.info.perpetuals.getClearinghouseState(this.walletAddress)
    );
    const state = raw as unknown as {
      accountValue?: string;
      withdrawable?: string;
      assetPositions?: Array<{
        position: {
          coin?: string;
          entryPx?: string;
          szi?: string;
          liquidationPx?: string;
          leverage?: unknown;
        };
      }>;
    };

    let totalValue = parseFloat(String(state.accountValue ?? 0));
    const enrichedPositions: EnrichedPosition[] = [];

    for (const posWrap of state.assetPositions ?? []) {
      const pos = posWrap.position;
      const entryPx = parseFloat(String(pos.entryPx ?? 0));
      const size = parseFloat(String(pos.szi ?? 0));
      const side = size > 0 ? "long" : "short";
      const currentPx = entryPx && size ? await this.getCurrentPrice(String(pos.coin)) : 0;
      const pnl =
        side === "long"
          ? (currentPx - entryPx) * Math.abs(size)
          : (entryPx - currentPx) * Math.abs(size);
      enrichedPositions.push({
        coin: String(pos.coin),
        szi: size,
        entryPx,
        pnl,
        liquidationPx: pos.liquidationPx ? parseFloat(String(pos.liquidationPx)) : null,
        leverage: pos.leverage,
      });
    }

    const balance = parseFloat(String(state.withdrawable ?? 0));
    if (!totalValue) {
      totalValue = balance + enrichedPositions.reduce((sum, p) => sum + Math.max(p.pnl, 0), 0);
    }

    return { balance, total_value: totalValue, positions: enrichedPositions };
  }

  async getOpenInterest(asset: string): Promise<number | null> {
    try {
      const [meta, assetCtxs] = await this.getMetaAndCtxs();
      const assetIdx = meta.universe.findIndex((u) => u.name === asset);
      if (assetIdx >= 0 && assetIdx < assetCtxs.length) {
        const oi = assetCtxs[assetIdx]?.openInterest;
        return oi ? Number(parseFloat(String(oi)).toFixed(2)) : null;
      }
      return null;
    } catch (err) {
      logger.error({ err, asset }, "OI fetch error");
      return null;
    }
  }

  async getFundingRate(asset: string): Promise<number | null> {
    try {
      const [meta, assetCtxs] = await this.getMetaAndCtxs();
      const assetIdx = meta.universe.findIndex((u) => u.name === asset);
      if (assetIdx >= 0 && assetIdx < assetCtxs.length) {
        const funding = assetCtxs[assetIdx]?.funding;
        return funding ? Number(parseFloat(String(funding)).toFixed(8)) : null;
      }
      return null;
    } catch (err) {
      logger.error({ err, asset }, "Funding fetch error");
      return null;
    }
  }
}
